scrapy/http/response/text.py

```diff
@@ -82,7 +82,7 @@ def json(self):
         Deserialize a JSON document to a Python object.
         """
         if self._cached_decoded_json is _NONE:
-            self._cached_decoded_json = json.loads(self.text)
+            self._cached_decoded_json = json.loads(self.body)
```

Comment:
The behavior is not exactly the same, if the response is not UTF-8, and properly indicates it.

From https://datatracker.ietf.org/doc/html/rfc7159#section-8.1:

> JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32. 

From https://docs.python.org/3/library/json.html#json.loads:

> s can now be of type [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) or [bytearray](https://docs.python.org/3/library/stdtypes.html#bytearray). The input encoding should be UTF-8, UTF-16 or UTF-32.

So we should be fine in most cases after this change. 

At the same time, SHALL in the RFC is not MUST (see https://datatracker.ietf.org/doc/html/rfc2119#section-3):

> SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

I'm fine with making the change, but we should add an entry to the changelog, which explains that the change is backwards incompatible, and that Scrapy stops support some of these weird cases to improve the efficiency for the vast majority of cases.

Reply:
Done! I've just added an entry to the **Backward-incompatible changes** section in the changelog. Please let me know if I need to correct something.